File: src/otiocat.zig
//! otiocat - Universal timeline format converter
//!
//! Reads .otio (JSON), .ziggy, .tlb (CBOR binary), .tlfb (FlatBuffers binary),
//! or .tlz (ZIP bundle) files and writes to .ziggy, .tlb, .tlfb, or .tlz format
//! based on the output file extension.
//!
//! Usage:
//!   otiocat <input> [output]
//!
//! If no output file is specified, prints Ziggy format to stdout.
//!
//! Examples:
//!   otiocat timeline.otio                   # JSON to Ziggy (stdout)
//!   otiocat timeline.otio timeline.ziggy    # JSON to Ziggy (file)
//!   otiocat timeline.otio timeline.tlb      # JSON to Binary (CBOR)
//!   otiocat timeline.otio timeline.tlfb     # JSON to Binary (FlatBuffers)
//!   otiocat timeline.ziggy timeline.tlz     # Ziggy to TLZ bundle
//!   otiocat timeline.tlz timeline.ziggy     # TLZ bundle to Ziggy
//!   otiocat timeline.ziggy timeline.tlb     # Ziggy to Binary
//!   otiocat timeline.tlb timeline.ziggy     # Binary to Ziggy
//!   otiocat timeline.tlfb timeline.ziggy    # FlatBuffers to Ziggy

const std = @import("std");
const string = @import("string_stuff");
const otio = @import("opentimelineio");
const serialization = otio.serialization;
const tlz_bundle_utils = otio.tlz_bundle_utils;

const builtin = @import("builtin");

/// Re-export MetadataMode from serialization for argument parsing
const MetadataMode = serialization.MetadataMode;

const State = struct {
    input_path: []const u8,
    output_path: ?[]const u8,
    metadata_mode: MetadataMode = .hash_reference,

    // TLZ bundle options
    bundle_format: tlz_bundle_utils.BundleFormat = .ziggy,
    media_policy: tlz_bundle_utils.MediaReferencePolicy = .MissingIfNotFile,

    pub fn deinit(
        self: @This(),
        allocator: std.mem.Allocator,
    ) void
    {
        allocator.free(self.input_path);
        if (self.output_path) |path| allocator.free(path);
    }
};

/// Parse the commandline arguments and setup the state
fn parse_args(
    allocator: std.mem.Allocator,
) !State
{
    var args = try std.process.argsWithAllocator(allocator);
    defer args.deinit();

    var input_path: ?[]const u8 = null;
    var output_path: ?[]const u8 = null;
    var metadata_mode: MetadataMode = .hash_reference;
    var bundle_format: tlz_bundle_utils.BundleFormat = .ziggy;
    var media_policy: tlz_bundle_utils.MediaReferencePolicy = .MissingIfNotFile;

    // Ignore the app name, always first in args
    _ = args.skip();

    var positional_count: usize = 0;

    // Read all the arguments from the commandline
    while (args.next())
        |nextarg|
    {
        const arg: [:0]const u8 = nextarg;

        // Handle options (flags starting with -)
        if (string.eql_latin_s8(arg, "--help") or
            string.eql_latin_s8(arg, "-h"))
        {
            usage("");
        }
        else if (string.eql_latin_s8(arg, "--no-metadata"))
        {
            metadata_mode = .no_metadata;
        }
        else if (string.eql_latin_s8(arg, "--inline-metadata"))
        {
            metadata_mode = .inline_metadata;
        }
        // TLZ bundle format options
        else if (string.eql_latin_s8(arg, "--bundle-format=ziggy"))
        {
            bundle_format = .ziggy;
        }
        else if (string.eql_latin_s8(arg, "--bundle-format=tlfb"))
        {
            bundle_format = .tlfb;
        }
        // TLZ media policy options
        else if (string.eql_latin_s8(arg, "--media-policy=error"))
        {
            media_policy = .ErrorIfNotFile;
        }
        else if (string.eql_latin_s8(arg, "--media-policy=missing"))
        {
            media_policy = .MissingIfNotFile;
        }
        else if (string.eql_latin_s8(arg, "--media-policy=all-missing"))
        {
            media_policy = .AllMissing;
        }
        else if (arg.len > 0 and arg[0] == '-')
        {
            // Unknown flag
            usage("Unknown option.");
        }
        else
        {
            // Positional argument (file path)
            positional_count += 1;
            switch (positional_count) {
                1 => {
                    input_path = try allocator.dupe(u8, arg);
                },
                2 => {
                    output_path = try allocator.dupe(u8, arg);
                },
                else => {
                    usage("Too many arguments.");
                },
            }
        }
    }

    if (positional_count < 1)
    {
        usage("Not enough arguments.");
    }

    return .{
        .input_path = input_path.?,
        .output_path = output_path,
        .metadata_mode = metadata_mode,
        .bundle_format = bundle_format,
        .media_policy = media_policy,
    };
}

/// Usage message for argument parsing
pub fn usage(
    msg: []const u8,
) noreturn
{
    std.debug.print(
        \\
        \\otiocat - Universal timeline format converter
        \\
        \\Reads timeline files in various formats and converts them to other formats.
        \\
        \\Supported input formats:
        \\  .otio   OpenTimelineIO JSON format
        \\  .ziggy  Ziggy text format
        \\  .tlb    Binary CBOR format
        \\  .tlfb   Binary FlatBuffers format
        \\  .tlz    TLZ bundle (ZIP archive with timeline + media)
        \\
        \\Supported output formats:
        \\  .ziggy  Ziggy text format
        \\  .tlb    Binary CBOR format
        \\  .tlfb   Binary FlatBuffers format
        \\  .tlz    TLZ bundle (ZIP archive with timeline + media)
        \\
        \\Usage:
        \\  otiocat [options] <input> [output]
        \\
        \\Arguments:
        \\  <input>   Path to the source timeline file
        \\  [output]  Path for the converted output file (optional)
        \\            If omitted, prints Ziggy format to stdout.
        \\
        \\Options:
        \\  -h, --help         Print this message and exit
        \\  --no-metadata      Omit all metadata from Ziggy output
        \\  --inline-metadata  Print metadata inline on each clip instead of
        \\                     using hash references (Ziggy output only)
        \\
        \\TLZ Bundle Options (for .tlz output):
        \\  --bundle-format=ziggy   Use Ziggy text format inside bundle (default)
        \\  --bundle-format=tlfb    Use FlatBuffers binary format inside bundle
        \\  --media-policy=error    Error if any media file not found
        \\  --media-policy=missing  Skip missing media files (default)
        \\  --media-policy=all-missing  Don't bundle any media files
        \\
        \\Examples:
        \\  otiocat timeline.otio                   # JSON to Ziggy (stdout)
        \\  otiocat timeline.otio timeline.ziggy    # JSON to Ziggy (file)
        \\  otiocat timeline.otio timeline.tlb      # JSON to Binary (CBOR)
        \\  otiocat timeline.otio timeline.tlfb     # JSON to FlatBuffers
        \\  otiocat timeline.ziggy timeline.tlb     # Ziggy to Binary
        \\  otiocat timeline.tlb timeline.ziggy     # Binary to Ziggy
        \\  otiocat timeline.tlfb timeline.ziggy    # FlatBuffers to Ziggy
        \\
        \\  # TLZ bundle examples:
        \\  otiocat timeline.ziggy timeline.tlz     # Create TLZ bundle
        \\  otiocat timeline.tlz timeline.ziggy     # Extract from TLZ bundle
        \\  otiocat timeline.ziggy timeline.tlz --bundle-format=tlfb  # Binary inside
        \\  otiocat timeline.ziggy timeline.tlz --media-policy=error  # Require media
        \\
        \\  # Metadata options (Ziggy output only):
        \\  otiocat --no-metadata timeline.otio     # No metadata in output
        \\  otiocat --inline-metadata timeline.otio # Metadata inline on clips
        \\
        \\{s}
        , .{msg}
    );
    std.process.exit(1);
}

/// Get file extension from path
fn get_extension(
    path: []const u8,
) ?[]const u8
{
    const ext_start = std.mem.lastIndexOfScalar(u8, path, '.') orelse return null;
    return path[ext_start..];
}


pub fn main() !void
{
    // Use the debug allocator in debug builds, otherwise use smp
    const parent_allocator = if (builtin.mode == .Debug) alloc: {
        var da = std.heap.DebugAllocator(.{}){};
        break :alloc da.allocator();
    } else std.heap.smp_allocator;

    var arena = std.heap.ArenaAllocator.init(parent_allocator);
    const allocator = arena.allocator();
    defer arena.deinit();

    const state = try parse_args(allocator);
    defer state.deinit(allocator);

    const prog = std.Progress.start(.{});
    defer prog.end();

    const parent_prog = prog.start("Converting timeline", 3);

    // Check input file exists
    const read_prog = parent_prog.start("Reading input file...", 0);

    var found = true;
    std.fs.cwd().access(state.input_path, .{}) catch |e| switch (e) {
        error.FileNotFound => found = false,
        else => return e,
    };

    if (!found)
    {
        std.log.err(
            "File: {s} does not exist or is not accessible.",
            .{state.input_path}
        );
        std.process.exit(1);
    }

    // Check output extension (default to .ziggy for stdout)
    const output_ext = if (state.output_path) |path|
        get_extension(path) orelse {
            std.log.err("Output file must have an extension (.ziggy, .tlb, .tlfb, or .tlz)", .{});
            std.process.exit(1);
        }
    else
        ".ziggy";

    const output_format = std.meta.stringToEnum(
        serialization.FileFormat,
        output_ext[1..],  // Skip the leading dot
    ) orelse {
        std.log.err(
            "Unsupported output format: {s}. Use .ziggy, .tlb, .tlfb, or .tlz",
            .{output_ext}
        );
        std.process.exit(1);
    };

    // TLZ output requires an output file (can't write to stdout)
    if (output_format == .tlz and state.output_path == null)
    {
        std.log.err("TLZ output requires an output file path", .{});
        std.process.exit(1);
    }

    // Read input file using centralized reader (supports .otio, .ziggy, .tlb, .tlfb, .tlz)
    const ser_timeline = try serialization.read_from_file(allocator, state.input_path);

    read_prog.end();

    const convert_prog = parent_prog.start("Converting...", 0);

    // Build write options from state
    const input_dir = std.fs.path.dirname(state.input_path) orelse ".";
    const write_options = serialization.WriteOptions{
        .metadata_mode = state.metadata_mode,
        .bundle_format = state.bundle_format,
        .media_policy = state.media_policy,
        .media_base_dir = input_dir,
    };

    // Write output using centralized writer
    if (state.output_path) |path|
    {
        // Write to file using centralized function
        try serialization.write_to_file(allocator, ser_timeline, path, write_options);
    }
    else
    {
        // Write to stdout
        var out_file = std.fs.File.stdout();
        var file_writer_buffer: [16 * 1024]u8 = undefined;
        var file_writer = out_file.writer(&file_writer_buffer);
        const writer = &file_writer.interface;

        try serialization.write_to_writer(
            allocator,
            ser_timeline,
            output_ext,
            write_options,
            writer,
        );

        try writer.flush();
    }

    convert_prog.end();

    if (state.output_path) |path|
    {
        std.log.info("Wrote: {s}", .{path});
    }
}

File: src/cpp_examples/otiocat.cpp
// otiocat.cpp - C++ Implementation
//
// Universal timeline format converter - reads and displays timeline files.
// Note: Write functionality requires C API extensions not yet implemented.

#include <iostream>
#include <string>
#include <algorithm>
#include "opentimelineio.hpp"

void print_usage() {
    std::cerr << R"(
otiocat_cpp - Universal timeline format viewer/converter

Reads timeline files in various formats and displays information about them.

Supported input formats:
  .otio   OpenTimelineIO JSON format
  .ziggy  Ziggy text format
  .tlb    Binary FlatBuffers based format

Usage:
  otiocat_cpp [options] <input>

Arguments:
  <input>   Path to the source timeline file

Options:
  -h, --help         Print this message and exit
  -v, --verbose      Show detailed timeline information

Examples:
  otiocat_cpp timeline.otio           # Display timeline info
  otiocat_cpp -v timeline.otio        # Verbose output
)";
}

std::string get_extension(const std::string& path) {
    auto pos = path.rfind('.');
    if (pos == std::string::npos) return "";
    std::string ext = path.substr(pos);
    std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);
    return ext;
}

void print_timeline_summary(const otio::Timeline& tl, bool verbose) {
    std::cout << "Timeline: " << tl.name().value_or("(unnamed)") << "\n";
    std::cout << "  Tracks: " << tl.track_count() << "\n";

    size_t total_clips = 0;
    size_t total_gaps = 0;
    size_t total_transitions = 0;

    for (size_t i = 0; i < tl.track_count(); ++i) {
        auto track = tl.track_at(i);

        if (verbose) {
            std::cout << "\n  Track " << i << ": "
                      << track.name().value_or("(unnamed)") << "\n";
            auto bounds = track.bounds();
            if (bounds) {
                std::cout << "    Bounds: " << bounds->start.as_float() << "s - "
                          << bounds->end.as_float() << "s\n";
            }
            std::cout << "    Children: " << track.child_count() << "\n";
        }

        for (size_t j = 0; j < track.child_count(); ++j) {
            auto child = track.child_at(j);
            switch (child.type()) {
                case otio::ComposableType::Clip:
                    total_clips++;
                    if (verbose) {
                        auto* clip = child.as_clip();
                        std::cout << "      Clip: " << clip->name().value_or("(unnamed)");
                        auto bounds = clip->bounds();
                        if (bounds) {
                            std::cout << " [" << bounds->start.as_float() << "s - "
                                      << bounds->end.as_float() << "s]";
                        }
                        auto media = clip->media();
                        if (media.is_uri()) {
                            std::cout << "\n        Media: " << media.target_uri;
                        }
                        std::cout << "\n";
                    }
                    break;
                case otio::ComposableType::Gap:
                    total_gaps++;
                    if (verbose) {
                        auto* gap = child.as_gap();
                        std::cout << "      Gap: " << gap->name().value_or("(unnamed)");
                        auto bounds = gap->bounds();
                        if (bounds) {
                            std::cout << " [" << bounds->start.as_float() << "s - "
                                      << bounds->end.as_float() << "s]";
                        }
                        std::cout << "\n";
                    }
                    break;
                case otio::ComposableType::Transition:
                    total_transitions++;
                    if (verbose) {
                        auto* trans = child.as_transition();
                        std::cout << "      Transition: " << trans->name().value_or("(unnamed)")
                                  << " (kind: " << trans->kind() << ")\n";
                    }
                    break;
                default:
                    break;
            }
        }
    }

    std::cout << "\nSummary:\n";
    std::cout << "  Total clips: " << total_clips << "\n";
    std::cout << "  Total gaps: " << total_gaps << "\n";
    std::cout << "  Total transitions: " << total_transitions << "\n";
}

int main(int argc, char* argv[]) {
    std::string input_path;
    bool verbose = false;

    // Parse arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "-h" || arg == "--help") {
            print_usage();
            return 0;
        } else if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (arg[0] == '-') {
            std::cerr << "Unknown option: " << arg << "\n";
            print_usage();
            return 1;
        } else if (input_path.empty()) {
            input_path = arg;
        } else {
            std::cerr << "Too many arguments.\n";
            print_usage();
            return 1;
        }
    }

    if (input_path.empty()) {
        std::cerr << "Error: No input file specified.\n";
        print_usage();
        return 1;
    }

    // @TODO: add .tlc support?

    // Validate extension
    std::string input_ext = get_extension(input_path);
    if (
            input_ext != ".otio" 
            && input_ext != ".tlb" 
            && input_ext != ".tla"
            && input_ext != ".tlz"
    ) 
    {
        std::cerr << "Error: Unsupported input format: " << input_ext << "\n";
        std::cerr << "Use .otio, .tla, .tlb, or .tlz \n";
        return 1;
    }

    try {
        std::cout << "Reading: " << input_path << "\n\n";
        otio::Timeline tl = otio::read_from_file(input_path);

        print_timeline_summary(tl, verbose);

        return 0;

    } catch (const otio::FileNotFoundError& e) {
        std::cerr << "Error: File not found: " << input_path << "\n";
        return 1;
    } catch (const otio::ParseError& e) {
        std::cerr << "Error: Failed to parse " << input_path << ": "
                  << e.what() << "\n";
        return 1;
    } catch (const otio::OtioError& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}
